api/api_test.go

```diff
@@ -0,0 +1,98 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	url "net/url"
+
+	"github.com/SkynetLabs/blocker/database"
+	"github.com/SkynetLabs/blocker/skyd"
+	"github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// apiTester is a helper struct wrapping handlers of the underlying API that
+// record a certain request and parse the API response.
+type apiTester struct {
+	staticAPI *API
+}
+
+// newAPITester returns a new instance of apiTester
+func newAPITester(api *API) *apiTester {
+	return &apiTester{staticAPI: api}
+}
+
+// newTestAPI returns a new API instance
+func newTestAPI(dbName string, skyd skyd.API) (*API, error) {
+	// create a nil logger
+	logger := logrus.New()
+	logger.Out = ioutil.Discard
+
+	// create database
+	db, err := database.NewCustomDB(context.Background(), "mongodb://localhost:37017", dbName, options.Credential{
+		Username: "admin",
+		Password: "aO4tV5tC1oU3oQ7u",
+	}, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), database.MongoDefaultTimeout)
+	defer cancel()
+
+	// purge the database
+	err = db.Purge(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	// create the API
+	api, err := New(skyd, db, logger)
+	if err != nil {
+		return nil, err
+	}
+	return api, nil
+}
+
+// blocklistGET records an api call to GET /blocklist on the underlying API
+// using the given parameters and returns a parsed response.
+func (at *apiTester) blocklistGET(sort *string, offset, limit *int) (BlocklistGET, error) {
```

Comment:
Usually, we would return the entries in ascending fashion. So entries would be returned from old->newest.
The idea is to request them from newest->oldest and by doing so be able to quickly abrupt the paging.

The API returns all non-invalid entries. So if for whatever reason the "target" blocker is not effectively blocking anything, the blocker that is syncing with the target is going to block those entries. If an entry is deemed `invalid` it's only because the reported hash or skylink is deemed invalid, which I figured is something universal and thus should be excluded from the blocklist.

Reply:
Are you proposing to remove the `sort` parameter? I don't mind it's just that usually, we return in `ascending` fashion while in this case, we would want to fetch entries in `desc` fashion to get the latest additions. That's why I added it the parameter.

Reply:
agreed in 1:1 discord it's fine